File: clients/java/src/test/java/io/volant/E2ETest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;

/**
 * Optional live broker e2e. Skipped unless {@code VOLANT_E2E=1}.
 *
 * <pre>
 * cargo build -p volant-server
 * VOLANT_E2E=1 mvn -q test
 * </pre>
 */
@EnabledIfEnvironmentVariable(named = "VOLANT_E2E", matches = "1")
class E2ETest {
    private static Process proc;
    private static Path dataDir;
    private static String host = "127.0.0.1";
    private static int port;

    @BeforeAll
    static void startBroker() throws Exception {
        assumeTrue("1".equals(System.getenv("VOLANT_E2E")), "set VOLANT_E2E=1 to run live broker e2e");
        String existing = System.getenv("VOLANT_BROKER");
        if (existing != null && !existing.isEmpty()) {
            String[] parts = splitHostPort(existing);
            host = parts[0];
            port = Integer.parseInt(parts[1]);
            waitPort(host, port, 5_000);
            return;
        }
        Path binary = ensureServerBin();
        assumeTrue(binary != null, "volant-server not found; build with `cargo build -p volant-server` "
                + "or set VOLANT_SERVER / VOLANT_BROKER");
        dataDir = Files.createTempDirectory("volant-java-e2e-");
        port = freePort();
        proc = new ProcessBuilder(
                        binary.toString(),
                        "--listen",
                        host + ":" + port,
                        "--data-dir",
                        dataDir.toString())
                .directory(repoRoot().toFile())
                .redirectOutput(ProcessBuilder.Redirect.DISCARD)
                .redirectError(ProcessBuilder.Redirect.DISCARD)
                .start();
        try {
            waitPort(host, port, 15_000);
        } catch (Exception e) {
            stopBroker();
            throw e;
        }
    }

    @AfterAll
    static void stopBroker() {
        if (proc != null) {
            proc.destroy();
            try {
                if (!proc.waitFor(5, TimeUnit.SECONDS)) {
                    proc.destroyForcibly();
                    proc.waitFor(5, TimeUnit.SECONDS);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                proc.destroyForcibly();
            }
            proc = null;
        }
        if (dataDir != null) {
            try {
                deleteRecursively(dataDir);
            } catch (IOException ignored) {
                // best-effort
            }
        }
    }

    @Test
    void createProduceFetchMetadata() {
        String topic = "java-e2e-" + ProcessHandle.current().pid() + "-" + System.nanoTime();
        try (Client c = Client.connect(host, port, 5_000)) {
            int topicId = c.createTopic(topic, 1);
            assertTrue(topicId >= 1, "topic id=" + topicId);

            long off = c.produce(topic, 0, null, "hello".getBytes(StandardCharsets.UTF_8));
            assertEquals(0L, off);

            List<Record> recs = c.fetch(topic, 0, 0);
            assertEquals(1, recs.size());
            assertEquals(0L, recs.get(0).offset);
            assertNull(recs.get(0).key);
            assertArrayEquals("hello".getBytes(StandardCharsets.UTF_8), recs.get(0).value);

            Metadata meta = c.metadata();
            assertFalse(meta.brokers.isEmpty());
            assertTrue(meta.topics.stream().anyMatch(t -> topic.equals(t.name)), "topic missing from metadata");

            c.deleteTopic(topic);
            Metadata meta2 = c.metadata();
            assertTrue(meta2.topics.stream().noneMatch(t -> topic.equals(t.name)), "topic still present after delete");
        }
    }

    @Test
    void offsetCommitFetch() {
        String topic = "java-off-" + ProcessHandle.current().pid() + "-" + System.nanoTime();
        String group = "java-og-" + ProcessHandle.current().pid();
        try (Client c = Client.connect(host, port, 5_000)) {
            c.createTopic(topic, 1);
            long produced = c.produce(topic, 0, null, "hello".getBytes(StandardCharsets.UTF_8));
            assertEquals(0L, produced);
            c.offsetCommit(group, topic, 0, 5);
            List<Offset> offs = c.offsetFetch(group, topic);
            assertEquals(1, offs.size());
            assertEquals(0, offs.get(0).partition);
            assertEquals(5L, offs.get(0).offset);
            c.deleteTopic(topic);
        }
    }

    @Test
    void groupConsumerPollCommitResume() {
        String topic = "java-gc-" + ProcessHandle.current().pid() + "-" + System.nanoTime();
        String group = "java-gcg-" + ProcessHandle.current().pid();
        try (Client c = Client.connect(host, port, 5_000)) {
            c.createTopic(topic, 1);
            c.produce(topic, 0, null, "hello".getBytes(StandardCharsets.UTF_8));

            try (GroupConsumer g = GroupConsumer.join(c, group, List.of(topic), 10_000)) {
                assertFalse(g.memberId().isEmpty());
                assertTrue(g.generation() >= 1);
                assertEquals(1, g.assignment().size());
                assertEquals(topic, g.assignment().get(0).topic);
                List<Record> recs = new ArrayList<>();
                for (int i = 0; i < 8 && recs.isEmpty(); i++) {
                    recs.addAll(g.poll(200));
                }
                assertEquals(1, recs.size(), "expected the produced record");
                assertEquals(0L, recs.get(0).offset);
                g.commit();
            }

            List<Offset> offs = c.offsetFetch(group, topic);
            assertEquals(1, offs.size());
            assertEquals(1L, offs.get(0).offset);

            c.produce(topic, 0, null, "world".getBytes(StandardCharsets.UTF_8));
            try (GroupConsumer g = GroupConsumer.join(c, group, List.of(topic), 10_000)) {
                List<Record> recs = new ArrayList<>();
                for (int i = 0; i < 8 && recs.isEmpty(); i++) {
                    recs.addAll(g.poll(200));
                }
                assertEquals(1, recs.size(), "resume should see only the new record");
                assertEquals(1L, recs.get(0).offset);
                assertArrayEquals("world".getBytes(StandardCharsets.UTF_8), recs.get(0).value);
                g.commit();
            }
            c.deleteTopic(topic);
        }
    }

    @Test
    void groupConsumerStaticMembership() {
        String topic = "java-static-" + ProcessHandle.current().pid() + "-" + System.nanoTime();
        String group = "java-staticg-" + ProcessHandle.current().pid();
        try (Client c = Client.connect(host, port, 5_000)) {
            c.createTopic(topic, 1);
            try (GroupConsumer g = GroupConsumer.joinStatic(c, group, List.of(topic), 10_000, "inst-1")) {
                assertEquals("inst-1", g.groupInstanceId());
                assertEquals("static:inst-1", g.memberId());
            }
            c.deleteTopic(topic);
        }
    }

    @Test
    void joinHeartbeatLeave() {
        String topic = "java-grp-" + ProcessHandle.current().pid() + "-" + System.nanoTime();
        String group = "java-cg-" + ProcessHandle.current().pid();
        try (Client c = Client.connect(host, port, 5_000)) {
            c.createTopic(topic, 1);
            JoinGroupResult j = c.joinGroup(group, List.of(topic), 10000);
            assertFalse(j.memberId.isEmpty(), "expected broker-assigned member id");
            assertTrue(j.generation >= 1, "generation=" + j.generation);
            assertEquals(1, j.assignment.size());
            assertEquals(topic, j.assignment.get(0).topic);
            assertEquals(0, j.assignment.get(0).partition);
            c.heartbeat(group, j.memberId, j.generation);
            c.leaveGroup(group, j.memberId);
            c.deleteTopic(topic);
        }
    }

    private static Path repoRoot() {
        Path dir = Paths.get("").toAbsolutePath();
        for (int i = 0; i < 8 && dir != null; i++) {
            if (Files.isRegularFile(dir.resolve("Cargo.toml"))
                    && Files.isDirectory(dir.resolve("clients"))) {
                return dir;
            }
            dir = dir.getParent();
        }
        // src/test/java/io/volant → 6 parents = clients/java
        Path fromCwd = Paths.get("").toAbsolutePath();
        Path guess = fromCwd.resolve("../..").normalize();
        return Files.isRegularFile(guess.resolve("Cargo.toml")) ? guess : fromCwd;
    }

    private static Path findServerBin() {
        String env = System.getenv("VOLANT_SERVER");
        if (env != null && !env.isEmpty()) {
            Path p = Paths.get(env);
            if (Files.isRegularFile(p)) {
                return p;
            }
        }
        Path root = repoRoot();
        Path debug = root.resolve("target").resolve("debug").resolve("volant-server");
        if (Files.isRegularFile(debug)) {
            return debug;
        }
        Path release = root.resolve("target").resolve("release").resolve("volant-server");
        if (Files.isRegularFile(release)) {
            return release;
        }
        return null;
    }

    private static Path ensureServerBin() {
        Path found = findServerBin();
        if (found != null) {
            return found;
        }
        Path cargo = which("cargo");
        if (cargo == null) {
            return null;
        }
        try {
            Process build = new ProcessBuilder(cargo.toString(), "build", "-p", "volant-server")
                    .directory(repoRoot().toFile())
                    .redirectOutput(ProcessBuilder.Redirect.DISCARD)
                    .redirectError(ProcessBuilder.Redirect.DISCARD)
                    .start();
            if (!build.waitFor(300, TimeUnit.SECONDS) || build.exitValue() != 0) {
                return null;
            }
        } catch (Exception e) {
            return null;
        }
        return findServerBin();
    }

    private static Path which(String name) {
        String path = System.getenv("PATH");
        if (path == null) {
            return null;
        }
        for (String dir : path.split(":")) {
            Path p = Paths.get(dir, name);
            if (Files.isRegularFile(p) && Files.isExecutable(p)) {
                return p;
            }
        }
        return null;
    }

    private static int freePort() throws IOException {
        try (ServerSocket s = new ServerSocket()) {
            s.setReuseAddress(true);
            s.bind(new InetSocketAddress("127.0.0.1", 0));
            return s.getLocalPort();
        }
    }

    private static void waitPort(String host, int port, long timeoutMs) throws Exception {
        long deadline = System.currentTimeMillis() + timeoutMs;
        Exception last = null;
        while (System.currentTimeMillis() < deadline) {
            try (Socket s = new Socket()) {
                s.connect(new InetSocketAddress(host, port), 250);
                return;
            } catch (IOException e) {
                last = e;
                Thread.sleep(50);
            }
        }
        throw new IOException("broker did not listen on " + host + ":" + port + ": " + last);
    }

    private static String[] splitHostPort(String addr) {
        int idx = addr.lastIndexOf(':');
        if (idx <= 0) {
            throw new IllegalArgumentException("invalid VOLANT_BROKER: " + addr);
        }
        return new String[] {addr.substring(0, idx), addr.substring(idx + 1)};
    }

    private static void deleteRecursively(Path root) throws IOException {
        if (!Files.exists(root)) {
            return;
        }
        Files.walk(root)
                .sorted((a, b) -> b.compareTo(a))
                .forEach(p -> {
                    try {
                        Files.deleteIfExists(p);
                    } catch (IOException ignored) {
                        // best-effort
                    }
                });
    }
}

File: clients/go/e2e_test.go
package volant_test

import (
	"fmt"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
)

func TestE2ECreateProduceFetchMetadata(t *testing.T) {
	if os.Getenv("VOLANT_E2E") != "1" {
		t.Skip("set VOLANT_E2E=1 to run live broker e2e")
	}
	addr, cleanup := startBroker(t)
	defer cleanup()

	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	topic := fmt.Sprintf("go-e2e-%d-%d", os.Getpid(), time.Now().UnixNano())
	if err := c.CreateTopic(topic, 1); err != nil {
		t.Fatalf("CreateTopic: %v", err)
	}

	off, err := c.Produce(topic, 0, nil, []byte("hello"))
	if err != nil {
		t.Fatalf("Produce: %v", err)
	}
	if off != 0 {
		t.Fatalf("base offset=%d want 0", off)
	}

	recs, err := c.Fetch(topic, 0, 0)
	if err != nil {
		t.Fatalf("Fetch: %v", err)
	}
	if len(recs) != 1 {
		t.Fatalf("len(recs)=%d want 1", len(recs))
	}
	if recs[0].Offset != 0 || recs[0].Key != nil || string(recs[0].Value) != "hello" {
		t.Fatalf("record %+v", recs[0])
	}

	meta, err := c.Metadata()
	if err != nil {
		t.Fatalf("Metadata: %v", err)
	}
	found := false
	for _, tp := range meta.Topics {
		if tp.Name == topic {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("topic %q missing from metadata", topic)
	}
	if len(meta.Brokers) == 0 {
		t.Fatal("expected at least one broker")
	}

	if err := c.DeleteTopic(topic); err != nil {
		t.Fatalf("DeleteTopic: %v", err)
	}
	meta2, err := c.Metadata()
	if err != nil {
		t.Fatalf("Metadata after delete: %v", err)
	}
	for _, tp := range meta2.Topics {
		if tp.Name == topic {
			t.Fatalf("topic %q still present after delete", topic)
		}
	}
}

func TestE2EOffsetCommitFetch(t *testing.T) {
	if os.Getenv("VOLANT_E2E") != "1" {
		t.Skip("set VOLANT_E2E=1 to run live broker e2e")
	}
	addr, cleanup := startBroker(t)
	defer cleanup()

	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	topic := fmt.Sprintf("go-off-%d-%d", os.Getpid(), time.Now().UnixNano())
	group := fmt.Sprintf("go-g-%d", os.Getpid())
	if err := c.CreateTopic(topic, 1); err != nil {
		t.Fatalf("CreateTopic: %v", err)
	}
	if _, err := c.Produce(topic, 0, nil, []byte("hello")); err != nil {
		t.Fatalf("Produce: %v", err)
	}
	if err := c.OffsetCommit(group, topic, 0, 5); err != nil {
		t.Fatalf("OffsetCommit: %v", err)
	}
	offs, err := c.OffsetFetch(group, topic)
	if err != nil {
		t.Fatalf("OffsetFetch: %v", err)
	}
	if len(offs) != 1 || offs[0].Partition != 0 || offs[0].Offset != 5 {
		t.Fatalf("offsets %+v want [{0 5}]", offs)
	}
	if err := c.DeleteTopic(topic); err != nil {
		t.Fatalf("DeleteTopic: %v", err)
	}
}

func TestE2EJoinHeartbeatLeave(t *testing.T) {
	if os.Getenv("VOLANT_E2E") != "1" {
		t.Skip("set VOLANT_E2E=1 to run live broker e2e")
	}
	addr, cleanup := startBroker(t)
	defer cleanup()

	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	topic := fmt.Sprintf("go-grp-%d-%d", os.Getpid(), time.Now().UnixNano())
	group := fmt.Sprintf("go-cg-%d", os.Getpid())
	if err := c.CreateTopic(topic, 1); err != nil {
		t.Fatalf("CreateTopic: %v", err)
	}
	j, err := c.JoinGroup(group, []string{topic}, 10000)
	if err != nil {
		t.Fatalf("JoinGroup: %v", err)
	}
	if j.MemberID == "" {
		t.Fatal("expected broker-assigned member id")
	}
	if j.Generation < 1 {
		t.Fatalf("generation=%d want >= 1", j.Generation)
	}
	if len(j.Assignment) != 1 || j.Assignment[0].Topic != topic || j.Assignment[0].Partition != 0 {
		t.Fatalf("assignment %+v want [{%s 0}]", j.Assignment, topic)
	}
	if err := c.Heartbeat(group, j.MemberID, j.Generation); err != nil {
		t.Fatalf("Heartbeat: %v", err)
	}
	if err := c.LeaveGroup(group, j.MemberID); err != nil {
		t.Fatalf("LeaveGroup: %v", err)
	}
	if err := c.DeleteTopic(topic); err != nil {
		t.Fatalf("DeleteTopic: %v", err)
	}
}

func startBroker(t *testing.T) (addr string, cleanup func()) {
	t.Helper()
	if existing := os.Getenv("VOLANT_BROKER"); existing != "" {
		host, portS, err := net.SplitHostPort(existing)
		if err != nil {
			t.Fatalf("VOLANT_BROKER: %v", err)
		}
		port, err := strconv.Atoi(portS)
		if err != nil {
			t.Fatalf("VOLANT_BROKER port: %v", err)
		}
		if err := waitPort(host, port, 5*time.Second); err != nil {
			t.Fatal(err)
		}
		return existing, func() {}
	}

	bin := ensureServerBin(t)
	dir := t.TempDir()
	port := freePort(t)
	addr = fmt.Sprintf("127.0.0.1:%d", port)
	cmd := exec.Command(bin, "--listen", addr, "--data-dir", dir)
	cmd.Dir = repoRoot()
	cmd.Stdout = nil
	cmd.Stderr = nil
	if err := cmd.Start(); err != nil {
		t.Fatalf("start volant-server: %v", err)
	}
	if err := waitPort("127.0.0.1", port, 15*time.Second); err != nil {
		_ = cmd.Process.Kill()
		_, _ = cmd.Process.Wait()
		t.Fatal(err)
	}
	return addr, func() {
		_ = cmd.Process.Signal(os.Interrupt)
		done := make(chan struct{})
		go func() {
			_, _ = cmd.Process.Wait()
			close(done)
		}()
		select {
		case <-done:
		case <-time.After(5 * time.Second):
			_ = cmd.Process.Kill()
			_, _ = cmd.Process.Wait()
		}
	}
}

func repoRoot() string {
	_, file, _, _ := runtime.Caller(0)
	return filepath.Clean(filepath.Join(filepath.Dir(file), "..", ".."))
}

func findServerBin() string {
	if env := os.Getenv("VOLANT_SERVER"); env != "" {
		if st, err := os.Stat(env); err == nil && !st.IsDir() {
			return env
		}
	}
	root := repoRoot()
	for _, rel := range []string{
		filepath.Join("target", "debug", "volant-server"),
		filepath.Join("target", "release", "volant-server"),
	} {
		p := filepath.Join(root, rel)
		if st, err := os.Stat(p); err == nil && !st.IsDir() {
			return p
		}
	}
	return ""
}

func ensureServerBin(t *testing.T) string {
	t.Helper()
	if found := findServerBin(); found != "" {
		return found
	}
	if _, err := exec.LookPath("cargo"); err != nil {
		t.Skip("volant-server not found; build with `cargo build -p volant-server` or set VOLANT_SERVER / VOLANT_BROKER")
	}
	cmd := exec.Command("cargo", "build", "-p", "volant-server")
	cmd.Dir = repoRoot()
	out, err := cmd.CombinedOutput()
	if err != nil {
		t.Skipf("cargo build -p volant-server failed: %v\n%s", err, out)
	}
	if found := findServerBin(); found != "" {
		return found
	}
	t.Skip("volant-server not found after cargo build; set VOLANT_SERVER / VOLANT_BROKER")
	return ""
}

func freePort(t *testing.T) int {
	t.Helper()
	l, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	port := l.Addr().(*net.TCPAddr).Port
	_ = l.Close()
	return port
}

func waitPort(host string, port int, timeout time.Duration) error {
	deadline := time.Now().Add(timeout)
	addr := net.JoinHostPort(host, strconv.Itoa(port))
	var last error
	for time.Now().Before(deadline) {
		conn, err := net.DialTimeout("tcp", addr, 250*time.Millisecond)
		if err == nil {
			_ = conn.Close()
			return nil
		}
		last = err
		time.Sleep(50 * time.Millisecond)
	}
	return fmt.Errorf("broker did not listen on %s: %v", addr, last)
}
